x/upgrade/abci.go

```diff
@@ -3,22 +3,34 @@ package upgrade
 import (
 	"fmt"
 
+	"github.com/spf13/viper"
 	abci "github.com/tendermint/tendermint/abci/types"
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
 // BeginBlock will check if there is a scheduled plan and if it is ready to be executed.
+// If skip upgrade flag is set, it will skip and clear the upgrade plan
 // If it is ready, it will execute it if the handler is installed, and panic/abort otherwise.
 // If the plan is not ready, it will ensure the handler is not registered too early (and abort otherwise).
 //
-// The prupose is to ensure the binary is switch EXACTLY at the desired block, and to allow
+// The purpose is to ensure the binary is switched EXACTLY at the desired block, and to allow
 // a migration to be executed if needed upon this switch (migration defined in the new binary)
 func BeginBlocker(k Keeper, ctx sdk.Context, _ abci.RequestBeginBlock) {
+
 	plan, found := k.GetUpgradePlan(ctx)
 	if !found {
 		return
 	}
+
+	skipUpgrade := viper.Get(FlagUnsafeSkipUpgrade).(bool)
```

Comment:
I don't like reading viper here, makes this core logic hard to test.
If reading viper could be moved into the Module, and just pass it as an arg to this function, I would be happier.

Reply:
I feel it's hard to read this flag elsewhere,as we're writing into BeginBlocker which makes it hard to pass the flag as arguments.
Any suggesstions here.?

Reply:
I would read it here: https://github.com/regen-network/cosmos-sdk/blob/regen-network/upgrade-module/x/upgrade/module.go#L127-L129

This is an untested shim, and can pass more arguments in. This actual `BeginBlocker` function has lots of business logic to test and should be unit-testable.

(I don't change the architecture at all, just move the call one level up)

Reply:
Got it,I felt it's bad to refactor BeginBlocker format against other modules.But if module.go/BeginBlock is an untested shim,it should totally be fine.
Will add test scripts accordingly to this.